### PR TITLE
KNN select-top-k variants

### PR DESCRIPTION
**Pull request by achirkin**
Integrate two new implementations for knn's `select_k` function.

### CODE REVIEW

**Review by tfeher — CHANGES_REQUESTED** [review]
Hi Artem, thanks for this PR! Although it is still a work in progress, I think it makes sense to give you an early feedback. Here are my first batch of comments.

File: cpp/include/raft/spatial/knn/detail/selection_faiss.cuh
Comment (achirkin) [reply cpp/include/raft/spatial/knn/detail/selection_faiss.cuh:169]: Yep

File: cpp/include/raft/spatial/knn/detail/ivf_flat/bitonic_sort.cuh
Comment (achirkin) [reply cpp/include/raft/spatial/knn/detail/ivf_flat/bitonic_sort.cuh:105]: I'm actually thinking, maybe I should remove this extra comparison and don't bother with stable sorting? I haven't yet evaluated how this affects the overall performance.

File: cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh
Comment (tfeher) [inline cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh:378]: If we did not have indices in the beginning, then do we really need to generate one?

File: cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh
Comment (tfeher) [inline cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh:372]: Please add a short description, like "Loads values in val_arr_".

File: cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh
Comment (tfeher) [inline cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh:396]: Please add a short description, like "Loads values in val_buf_".

File: cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh
Comment (tfeher) [inline cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh:145]: Is there any relationship between `capacity` and `k` that we should check?

File: cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh
Comment (tfeher) [inline cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh:444]: Is it assumed that the input is already sorted?

**Review by tfeher — CHANGES_REQUESTED** [review]
Hi Artem, here is a second batch of comments.

File: cpp/include/raft/spatial/knn/knn.hpp
Comment (achirkin) [reply cpp/include/raft/spatial/knn/knn.hpp]: Do you mean to add one more overload in `::detail` with a flipped boolean argument?

File: cpp/include/raft/spatial/knn/knn.cuh
Comment (achirkin) [reply cpp/include/raft/spatial/knn/knn.cuh:41]: Not sure, this PR mainly updates the `select_k` below. I only update the template parameter names so they look more alike the rest.

File: cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh
Comment (achirkin) [reply cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh]: Apparently, the loop is unrolled, which ensures the constant indexing of the arrays. When directly indexing using `buf_len_`, the array is placed in local memory instead of the registers, which makes the whole algorithm slower (~2x slower on my machine).

File: cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh
Comment (achirkin) [reply cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh:417]: Yes, for kNN we need to find out distances between which elements are the smallest in the end. Sometimes we don't need to pass indices in the first place, because `0..n-1` is implied (dense inputs).

File: cpp/include/raft/spatial/knn/detail/ivf_flat/bitonic_sort.cuh
Comment (achirkin) [reply cpp/include/raft/spatial/knn/detail/ivf_flat/bitonic_sort.cuh:72]: That's used to delay the evaluation of this template till the same time the function templates are being compiled, rather than to compile it together with the enclosing struct.

File: cpp/include/raft/spatial/knn/detail/ivf_flat/radix_topk.cuh
Comment (achirkin) [reply cpp/include/raft/spatial/knn/detail/ivf_flat/radix_topk.cuh:415]: (NB: I removed the other function since the review). The interface now is exactly the same as the public one and the rest implementations in the `detail` namespace. I guess, no need to repeat the same docs multiple times?..

File: cpp/include/raft/spatial/knn/detail/ivf_flat/radix_topk.cuh
Comment (achirkin) [reply cpp/include/raft/spatial/knn/detail/ivf_flat/radix_topk.cuh]: I guess, undefined behavior.

Comment (tfeher) [reply cpp/include/raft/spatial/knn/detail/ivf_flat/radix_topk.cuh]: I meant to ask what causes here UB?

File: cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh
Comment (tfeher) [reply cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh]: Ok, then let us keep the loop how it is.

File: cpp/include/raft/spatial/knn/detail/ivf_flat/radix_topk.cuh
Comment (achirkin) [reply cpp/include/raft/spatial/knn/detail/ivf_flat/radix_topk.cuh:49]: I've had a feeling that both versions are common. I guess, here we go from the most to least significant bits, because this way we can save a few passes since, as you said, we don't need to sort within the `k`-sized buckets.

File: cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh
Comment (achirkin) [reply cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh]: Yes, it's ensured in the calling code that `k <= capacity`, but we cannot use raft's `ASSERT` in the device code to check the value of `k` at runtime.

**Review by tfeher — CHANGES_REQUESTED** [review]
Hi Artem, I have finished my first review round. Overall it looks good, but as my comments indicate, some more documentation is required at certain places.

File: cpp/include/raft/spatial/knn/knn.hpp
Comment (tfeher) [reply cpp/include/raft/spatial/knn/knn.hpp]: No, I meant to have only one variant, that has the same (flipped) boolean arg.

File: cpp/bench/spatial/selection.cu
Comment (cjnolet) [inline cpp/bench/spatial/selection.cu:2]: New files should just contain current year

File: cpp/bench/spatial/selection.cu
Comment (cjnolet) [inline cpp/bench/spatial/selection.cu:35]: Since this is a new file, we should use consistent casing/naming of template arguments. I vote for `idx_t` and `key_t` to match other functions in RAFT's nearest neighbors API.

File: cpp/include/raft/spatial/knn/detail/ivf_flat/bitonic_sort.cuh
Comment (cjnolet) [inline cpp/include/raft/spatial/knn/detail/ivf_flat/bitonic_sort.cuh:170]: same here w/ templates. We should strive to be consistent. Though, if we want to forego this discussion / change to get this PR in quickly for 22.04, can you create an issue for it so we don't lose track?

File: cpp/include/raft/spatial/knn/knn.cuh
Comment (cjnolet) [reply cpp/include/raft/spatial/knn/knn.cuh:41]: @tfeher, yes it's correct because it's performing the same k-select it would perform over a tile of pairwise distances, but it's doing it over potentially multiple k-selected partitions `n_partsxnxk` to reduce them into a single set of size `nxk`.

File: cpp/include/raft/spatial/knn/knn.cuh
Comment (cjnolet) [inline cpp/include/raft/spatial/knn/knn.cuh:85]: love the expansion in the documentation here.

File: cpp/include/raft/spatial/knn/knn.cuh
Comment (cjnolet) [inline cpp/include/raft/spatial/knn/knn.cuh:77]: Can you expand here to note the implementations which would not contain sorted values here? It's often assumed the resulting neighbors will be sorted by distance. For example, HDBSCAN and RBC expect that the kth neighbor is the `neighbor at k`, which is used to compute the radius.

File: cpp/include/raft/spatial/knn/knn.cuh
Comment (cjnolet) [inline cpp/include/raft/spatial/knn/knn.cuh:86]: Can you add `device array` to all of the inputs/outputs which are on device? We should start doing this across the codebase as we're adding and updating docs.

**Review by cjnolet — CHANGES_REQUESTED** [review]
I've done an initial scan through the changes to provide some feedback while I do a more extensive review. 

If we are to still aim for 22.04 with this PR, we also need a PR to cuML to make sure nothing breaks since it's so close to release time.

File: cpp/bench/spatial/selection.cu
Comment (achirkin) [reply cpp/bench/spatial/selection.cu:35]: I don't mind your variant, but I changed this cause I thought we're trying now to stick to the style declared in [cpp/.clang-tidy](https://github.com/rapidsai/raft/blob/15eb20a4e90a2c9267dea7046dc9cb9cba3e678e/cpp/.clang-tidy#L211-L212). So I just follow the suggestions of clang-tidy to rename things.

Would you like me to change the template parameter naming here or the rules in .clang-tidy?

File: cpp/include/raft/spatial/knn/knn.cuh
Comment (achirkin) [reply cpp/include/raft/spatial/knn/knn.cuh:87]: Added to the description for the enum class.

File: cpp/include/raft/spatial/knn/detail/ivf_flat/radix_topk.cuh
Comment (yong-wang) [reply cpp/include/raft/spatial/knn/detail/ivf_flat/radix_topk.cuh:267]: update of the counter:
The k-th value must be in the `index`-th bucket, so we record it in `counter->bucket`.
The number of values in this bucket is `(cur - prev)` since `histogram` now actually contains prefix sum of the histogram. Hence the update of `counter->len`.
counter is used in the next pass, so we keep the len of current pass in `counter->previous_len`.

`prev` means the number of values in the first `index-1` buckets. These values are less than (when `greater`=false) the real k-th value, so must be in the top-k results. In the next pass, we need only to find the `(k - prev)`-th value fall in the `index`-th bucket. Hence the setting of `counter->k`.

File: cpp/include/raft/spatial/knn/detail/ivf_flat/radix_topk.cuh
Comment (yong-wang) [reply cpp/include/raft/spatial/knn/detail/ivf_flat/radix_topk.cuh:383]: This `if` branch is for another top-k implementation called "radix-select", which only sets counter->kth_value during the passes without writing top-k values to the result array `out`.
After all passes are done, the counter->kth_value is used as a threshold to filter out the real top-k values from original inputs. So counter->kth_value is just an intermediate value and is never passed back to the caller.

Since this implementation is removed, I think this `if` branch is not used anymore and should be deleted.

File: cpp/include/raft/spatial/knn/detail/ivf_flat/radix_topk.cuh
Comment (yong-wang) [reply cpp/include/raft/spatial/knn/detail/ivf_flat/radix_topk.cuh:289]: Yes, it's done by the condition `current_value < k` of the `for` loop.
At the end of the loop, `current_value` is set to the last prefix sum that has been computed. When it >= k, we have found the right bucket, don't need scan further, and we break the loop.

Sorry, the name `current_value` is too vague to reason about its usage, maybe change it to `last_prefix_sum`?

Comment (yong-wang) [reply cpp/include/raft/spatial/knn/detail/ivf_flat/radix_topk.cuh]: It's done by the condition `current_value < k` of the for loop.
At the end of the loop, current_value is set to the last prefix sum that has been computed. When it >= k, we have found the right bucket, don't need scan further, and we break the loop.

Sorry, the name `current_value` is too vague to reason about its usage, maybe change it to `last_prefix_sum`?
(Re-post, I deleted this comment by mistake, sorry)

**Comment by cjnolet** [comment]
rerun tests

**Comment by cjnolet** [comment]
rerun tests

File: cpp/include/raft/spatial/knn/detail/ivf_flat/bitonic_sort.cuh
Comment (achirkin) [reply cpp/include/raft/spatial/knn/detail/ivf_flat/bitonic_sort.cuh]: Closing this with the reference to the new issue https://github.com/rapidsai/raft/issues/591 .

File: cpp/bench/spatial/selection.cu
Comment (achirkin) [reply cpp/bench/spatial/selection.cu:35]: Closing this with the reference to the new issue https://github.com/rapidsai/raft/issues/591 .

**Review by tfeher — CHANGES_REQUESTED** [review]
Thanks Artem for addressing the issues and improving the documentation! I had  second look at the bitonic sort, here are a few additional comments.

File: cpp/include/raft/spatial/knn/detail/topk/bitonic_sort.cuh
Comment (achirkin) [reply cpp/include/raft/spatial/knn/detail/topk/bitonic_sort.cuh]: Yep, sounds good

File: cpp/include/raft/spatial/knn/detail/topk/radix_topk.cuh
Comment (tfeher) [inline cpp/include/raft/spatial/knn/detail/topk/radix_topk.cuh:224]: Is out always a valid pointer? Three lines above we do not check, but here we do. Please make it consistent.

**Review by tfeher — CHANGES_REQUESTED** [review]
Hi Artem, I had a second look on radix_topk. It looks good overall, just found a few minor things.

File: cpp/include/raft/spatial/knn/detail/topk/warpsort_topk.cuh
Comment (tfeher) [inline cpp/include/raft/spatial/knn/detail/topk/warpsort_topk.cuh:97]:
```suggestion
        queue.load(in, 0, len);
```

**Review by tfeher — CHANGES_REQUESTED** [review]
I have finished a second review. Please find a few more suggestions for improving the docs. Apart from these pending items, the PR looks good to me.

File: cpp/include/raft/spatial/knn/detail/topk/warpsort_topk.cuh
Comment (tfeher) [inline cpp/include/raft/spatial/knn/detail/topk/warpsort_topk.cuh:142]: Too much whitespace.

**Review by tfeher — APPROVED** [review]
Thanks Artem for addressing the issues! Apart from two more nitpicks, the PR looks good to me. Pre-approving.

File: cpp/include/raft/spatial/knn/detail/topk/warpsort_topk.cuh
Comment (achirkin) [reply cpp/include/raft/spatial/knn/detail/topk/warpsort_topk.cuh]: Sorry :)

File: cpp/bench/spatial/selection.cu
Comment (cjnolet) [inline cpp/bench/spatial/selection.cu:103]: Are these settings for `k` failing? Can we remove the commented out bits in the meantime?

**Review by cjnolet — CHANGES_REQUESTED** [review]
Thanks Artem! Just one little thing from me this time around. I don't see any obvious things in these changes that would break cuml downstream but a PR should be opened and pinned to this branch before merging just to make sure.

**Comment by cjnolet** [comment]
rerun tests

File: cpp/bench/spatial/selection.cu
Comment (achirkin) [reply cpp/bench/spatial/selection.cu]: Yes, since only the radix algorithm supports larger values of `k`.

**Review by cjnolet — APPROVED** [review]
LGTM

**Comment by cjnolet** [comment]
@gpucibot merge

**Pull request merged**
